src/DotNetty.Transport/Channels/Sockets/SocketDatagramChannel.cs

```diff
@@ -0,0 +1,479 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace DotNetty.Transport.Channels.Sockets
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.Contracts;
+    using System.Net;
+    using System.Net.NetworkInformation;
+    using System.Net.Sockets;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using DotNetty.Buffers;
+    using DotNetty.Common;
+    using DotNetty.Common.Concurrency;
+    using DotNetty.Common.Internal.Logging;
+    using DotNetty.Common.Utilities;
+
+    public class SocketDatagramChannel : AbstractSocketMessageChannel, ISocketDatagramChannel
+    {
+        static readonly IInternalLogger Logger = InternalLoggerFactory.GetInstance<SocketDatagramChannel>();
+        static readonly Action<object, object> ReceiveFromCompletedSyncCallback = OnReceiveFromCompletedSync;
+        static readonly ChannelMetadata ChannelMetadata = new ChannelMetadata(true);
+
+        readonly DefaultSocketDatagramChannelConfig config;
+        readonly IPEndPoint anyRemoteEndPoint;
+
+        public SocketDatagramChannel()
+            : this(new Socket(SocketType.Dgram, ProtocolType.Udp))
+        {
+        }
+
+        public SocketDatagramChannel(AddressFamily addressFamily)
+            : this(new Socket(addressFamily, SocketType.Dgram, ProtocolType.Udp))
+        {
+        }
+
+        public SocketDatagramChannel(Socket socket)
+            : base(null, socket)
+        {
+            this.config = new DefaultSocketDatagramChannelConfig(this, socket);
+            this.anyRemoteEndPoint = new IPEndPoint(
+                socket.AddressFamily == AddressFamily.InterNetwork ? IPAddress.Any : IPAddress.IPv6Any,
+                IPEndPoint.MinPort);
+
+            // Read is not allowed until socket is bound for UDP
+            this.config.AutoRead = false;
```

Comment:
AutoRead is meant for user of DotNetty to control channel's behavior when it comes to reading data out of source. Here we take control away from the user. How is it handled in netty?